src/ts/index.ts

```diff
@@ -0,0 +1,296 @@
+   *   requiring a search.
+   */
+  public get tokens(): string[] {
+    // NOTE: Make sure to slice the original array, so it cannot be modified
+    // outside of this class.
+    return this._tokens.slice();
+  }
+
+  /**
+   * Gets the encoded orders as a hex-encoded string.
+   */
+  public get encodedOrders(): string {
+    return this._encodedOrders;
+  }
+
+  /**
+   * Gets the number of orders currently encoded.
+   */
+  public get orderCount(): number {
+    // NOTE: `ORDER_STRIDE` multiplied by 2 as hex strings encode one byte in
+    // 2 characters.
+    return (this._encodedOrders.length - 2) / (ORDER_STRIDE * 2);
+  }
+
+  /**
+   * Encodes a signed order, appending it to the `calldata` bytes that are being
+   * built.
+   *
+   * Additionally, if the order references new tokens that the encoder has not
+   * yet seen, they are added to the tokens array.
+   * @param order The order to encode.
+   * @param executedAmount The executed trade amount for the order.
+   * @param signature The signature for the order data.
+   */
+  public encodeOrder(
+    order: Order,
+    executedAmount: BigNumberish,
+    signature: SignatureLike,
+  ): void {
+    const sig = ethers.utils.splitSignature(signature);
+    const encodedOrder = ethers.utils.solidityPack(
+      [
+        "uint8",
+        "uint8",
+        "uint256",
+        "uint256",
+        "uint32",
+        "uint32",
+        "uint256",
+        "uint8",
+        "uint256",
+        "uint8",
+        "bytes32",
+        "bytes32",
+      ],
+      [
+        this.tokenIndex(order.sellToken),
+        this.tokenIndex(order.buyToken),
+        order.sellAmount,
+        order.buyAmount,
+        timestamp(order.validTo),
+        order.nonce,
+        order.tip,
+        encodeOrderFlags(order),
+        executedAmount,
+        sig.v,
+        sig.r,
+        sig.s,
+      ],
+    );
+
+    this._encodedOrders = `${this._encodedOrders}${encodedOrder.substr(2)}`;
+  }
+
+  /**
+   * Signs and encodes an order.
+   * @param owner The owner for the order used to sign.
+   * @param order The order to compute the digest for.
+   * @param executedAmount The executed trade amount for the order.
+   * @return Signature for the order.
+   */
+  public signEncodeOrder(
+    owner: Wallet,
+    order: Order,
+    executedAmount: BigNumberish,
+  ): void {
+    if (!this.domainSeparator) {
+      throw new Error("domain separator not specified");
+    }
+    const signature = signOrder(owner, this.domainSeparator, order);
+    this.encodeOrder(order, executedAmount, signature);
+  }
+
+  private tokenIndex(token: string): number {
+    // NOTE: Verify and normalize the address into a case-checksummed address.
+    // Not only does this ensure validity of the addresses early on, it also
+    // makes it so `0xff...f` and `0xFF..F` map to the same ID.
+    const tokenAddress = ethers.utils.getAddress(token);
+
+    let tokenIndex = this._tokenMap[tokenAddress];
```

Comment:
No need to change now, but I'm not sure it's a good idea to populate the token Map this implicitly given that any solution needs to also contain prices that are aligned with this list. 
Maybe setting the token vector explicitly from above (so that it can be colocated with the logic that extracts the prices vector from a solution) will become necessary.

Reply:
I thought that as well. I just kept it this way as it was intuitive for the test cases, but I also agree the interface will have to change to encode full settlements further down the road.